File: Web/backend/src/main/java/com/ByTech/ByTech/Fabricante/services/EnterpriseService.java
package com.ByTech.ByTech.Fabricante.services;

import com.ByTech.ByTech.Fabricante.models.EnterpriseModel;
import com.ByTech.ByTech.Fabricante.repositories.IEnterpriseRepository;
import com.ByTech.ByTech.User.models.UserModel;
import com.ByTech.ByTech.User.repositories.IUserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Optional;

@Service
public class EnterpriseService {

    @Autowired
    IEnterpriseRepository enterpriseRepository;
    @Autowired
    IUserRepository userRepository;

    public ArrayList<EnterpriseModel> getEnterprise() {
        return (ArrayList<EnterpriseModel>) enterpriseRepository.findAll();
    }

    public EnterpriseModel saveEnterprise(EnterpriseModel enterprise) throws Exception {
        if (existsByNombre(enterprise.getNombre())) {
            throw new Exception("Enterprise already exists");
        }
        return enterpriseRepository.save(enterprise);
    }

    public Optional<EnterpriseModel> getById(Long id) {
        return enterpriseRepository.findById(id);
    }

    public EnterpriseModel updateByID(EnterpriseModel request, Long id) throws Exception {
        EnterpriseModel enterpriseModel = enterpriseRepository.findById(id).orElseThrow(() -> new Exception("Enterprise not found"));

        if (!enterpriseModel.getNombre().equals(request.getNombre()) && existsByNombre(request.getNombre())) {
            throw new Exception("Enterprise already exists");
        }

        enterpriseModel.setNif(request.getNif());
        enterpriseModel.setNombre(request.getNombre());
        enterpriseModel.setBoss(request.getBoss());
        enterpriseModel.setDescripcion(request.getDescripcion());
        enterpriseModel.setWorkers(request.getWorkers());

        return enterpriseRepository.save(enterpriseModel);
    }

    public Boolean deleteEnterprise(Long id) {
        try {
            enterpriseRepository.deleteById(id);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public boolean existsByNombre(String nombre) {
        return enterpriseRepository.existsByNombre(nombre);
    }

}

File: Movil/backend/ByTech_movil/src/main/kotlin/com/example/ByTech_movil/Fabricante/services/EnterpriseService.kt
package com.example.ByTech_movil.Fabricante.services

import com.example.ByTech_movil.Fabricante.models.EnterpriseModel
import com.example.ByTech_movil.Fabricante.repositories.IEnterpriseRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.*

@Service
class EnterpriseService {
    @Autowired
    lateinit var enterpriseRepository: IEnterpriseRepository

    val enterprise: List<EnterpriseModel>
        get() = enterpriseRepository.findAll()

    @Throws(Exception::class)
    fun saveEnterprise(enterprise: EnterpriseModel): EnterpriseModel {
        if (existsByNombre(enterprise.nombre ?: "")) {
            throw Exception("Enterprise already exists")
        }
        return enterpriseRepository.save(enterprise)
    }

    fun getById(id: Long?): Optional<EnterpriseModel> {
        return enterpriseRepository.findById(id ?: -1)
    }

    @Throws(Exception::class)
    fun updateByID(request: EnterpriseModel, id: Long?): EnterpriseModel {
        val enterpriseModel: EnterpriseModel =
            enterpriseRepository.findById(id ?: -1).orElseThrow { Exception("Enterprise not found") }

        if (enterpriseModel.nombre != request.nombre && existsByNombre(request.nombre ?: "")) {
            throw Exception("Enterprise already exists")
        }

        enterpriseModel.nif = request.nif
        enterpriseModel.nombre = request.nombre
        enterpriseModel.boss = request.boss
        enterpriseModel.descripcion = request.descripcion
        enterpriseModel.workers = request.workers?.toMutableList()

        return enterpriseRepository.save(enterpriseModel)
    }

    fun deleteEnterprise(id: Long?): Boolean {
        return try {
            if (id != null && enterpriseRepository.existsById(id)) {
                enterpriseRepository.deleteById(id)
                true
            } else {
                false
            }
        } catch (e: Exception) {
            false
        }
    }

    fun existsByNombre(nombre: String): Boolean {
        return enterpriseRepository.existsByNombre(nombre)
    }
}
